DataFormats/TrackerRecHit2D/src/VectorHit.cc

```diff
@@ -0,0 +1,247 @@
+  return phase2clusterGlobalPos(geomDetUpper, upperCluster());
+}
+
+Global3DPoint VectorHit::phase2clusterGlobalPos(const PixelGeomDetUnit* geomDet, ClusterRef cluster) {
+  const PixelTopology* topo = &geomDet->specificTopology();
+  float ix = cluster->center();
+  float iy = cluster->column() + 0.5;                    // halfway the column
+  LocalPoint lp(topo->localX(ix), topo->localY(iy), 0);  // x, y, z
+  Global3DPoint gp = geomDet->surface().toGlobal(lp);
+  return gp;
+}
+
+GlobalError VectorHit::lowerGlobalPosErr() const {
+  const StackGeomDet* stackDet = dynamic_cast<const StackGeomDet*>(det());
+  const PixelGeomDetUnit* geomDetLower = dynamic_cast<const PixelGeomDetUnit*>(stackDet->lowerDet());
+  return phase2clusterGlobalPosErr(geomDetLower);
+}
+
+GlobalError VectorHit::upperGlobalPosErr() const {
+  const StackGeomDet* stackDet = dynamic_cast<const StackGeomDet*>(det());
+  const PixelGeomDetUnit* geomDetUpper = dynamic_cast<const PixelGeomDetUnit*>(stackDet->upperDet());
+  return phase2clusterGlobalPosErr(geomDetUpper);
+}
+
+GlobalError VectorHit::phase2clusterGlobalPosErr(const PixelGeomDetUnit* geomDet) {
+  const PixelTopology* topo = &geomDet->specificTopology();
+  float pitchX = topo->pitch().first;
+  float pitchY = topo->pitch().second;
+  constexpr float invTwelve = 1. / 12;
+  LocalError le(pow(pitchX, 2) * invTwelve, 0, pow(pitchY, 2) * invTwelve);  // e2_xx, e2_xy, e2_yy
+  GlobalError ge(ErrorFrameTransformer().transform(le, geomDet->surface()));
+  return ge;
+}
+
+Global3DVector VectorHit::globalDelta() const {
+  Local3DVector theLocalDelta =
+      LocalVector(theDirection.x() * theDirection.z(), theDirection.y() * theDirection.z(), theDirection.z());
+  Global3DVector g = det()->surface().toGlobal(theLocalDelta);
+  return g;
+}
+
+Global3DVector VectorHit::globalDirection() const { return (det()->surface().toGlobal(localDirection())); }
+
+float VectorHit::theta() const { return globalDirection().theta(); }
+
+float VectorHit::transverseMomentum(float magField) const {
+  return magField * 2.99792458e-3F / theCurvature;
+}  // pT [GeV] ~ 0.3 * B[T] * R [m], curvature is in cms, thus using 2.99792458e-3F (precise value from speed of light)
+float VectorHit::momentum(float magField) const { return transverseMomentum(magField) / (1. * sin(theta())); }
+
+AlgebraicMatrix VectorHit::projectionMatrix() const {
+  // obsolete (for what tracker is concerned...) interface
+  static const AlgebraicMatrix the4DProjectionMatrix(nComponents, 5, 0);
+  return the4DProjectionMatrix;
+}
+
+LocalError VectorHit::localPositionError() const {
+  return LocalError(theCovMatrix[2][2], theCovMatrix[2][3], theCovMatrix[3][3]);
+}
+
+LocalError VectorHit::localDirectionError() const {
+  return LocalError(theCovMatrix[0][0], theCovMatrix[0][1], theCovMatrix[1][1]);
+}
+
+AlgebraicSymMatrix VectorHit::parametersError() const {
+  //think about a more efficient method
```

Comment:
`AlgebraicSymMatrix` (which comes from CLHEP, https://github.com/cms-sw/cmssw/blob/master/DataFormats/CLHEP/interface/AlgebraicObjects.h#L15) has a copy constructor. Why does this function need to make a copy, fill it entry by entry, and then return that copy? Can't it just return the actual `theCovMatrix` object, and then the `VectorHit` copy constructor just copies it directly, once?

Reply:
I had noticed this too earlier and the loop is indeed unnecessary, in fact the igprof profile I linked to actually already uses the copy constructor. Unfortunately, that is the really slow part, the loop didn't contribute much to the overall cost.